File: marc/examples/nqueens/src/NQueens.cpp
#include "../header/NQueens.h"
#include "../header/QueensBoard.h"
#include <random>


double NQueens::heuristic(const std::shared_ptr<Modeling::State>& s) const {
    double res = 0.0;
    const unsigned int queens = dynamic_cast<const IntDimensionType&>(this->getDim()).getDim();
    // Note 1. Downcasting is needed here (general State type to this Problem's State type)
    auto board = std::dynamic_pointer_cast<QueensBoard>(s)->getValue(); 

    for (unsigned int i = 0; i < queens; i++) {
        for (unsigned int j = i+1; j < queens; j++) {
            if (board[i] == board[j] || (abs(board[j] - board[i]) == j - i)) res++;
        }
    }

    return res;

}

std::shared_ptr<Modeling::State> NQueens::getRandomState() const {
    const unsigned int queens = dynamic_cast<const IntDimensionType&>(this->getDim()).getDim();
    auto arr = std::make_unique<std::vector<unsigned int>>();
    std::random_device dev;
    std::mt19937 rng(dev());
    std::uniform_int_distribution<std::mt19937::result_type> dist(0, queens);

    for (unsigned int i = 0; i < queens; i++) arr->push_back(dist(rng));
    return std::make_shared<QueensBoard>(std::move(arr));
}

bool NQueens::objectiveTest(const std::shared_ptr<Modeling::State>& s) const {
    return this->heuristic(s) == 0; // Note 2. Objective Test is always trivial if heuristic is defined
}

NQueens::NQueens(const unsigned int dim) : 
    Modeling::RandomExplorationProblem(std::make_unique<IntDimensionType>(dim)),
    Modeling::HeuristicDrivenSearchProblem(std::make_unique<IntDimensionType>(dim)),
    Modeling::InitialStateProblem(std::make_unique<IntDimensionType>(dim)),
    Modeling::Problem(std::make_unique<IntDimensionType>(dim)) { }

File: jmarc/examples/NQueens/NQueens.java
import jmarc.modeling.*;

import java.util.Random;
import java.util.ArrayList;


public class NQueens extends MARCInitStateProblem<Integer> implements MARCHeuristicProblem, MARCRandomProblem {

    private Integer queens;

    public boolean objectiveTest(MARCState s) {
        return this.heuristic(s) == 0;
    }

    public double heuristic(MARCState s) {
        double res = 0;
        ArrayList<Integer> board = ((QueensBoard)s).getValue();

        for (int i = 0; i < queens; i++) {
            for (int j = i+1; j < queens; j++) {
                if (board.get(i).equals(board.get(j)) || (Math.abs(board.get(j) - board.get(i)) == j - i)) res++;
            }
        }

        return res;

    }


    private int countConflicts(int index, int newVal, ArrayList<Integer> board) {
        int res = 0;
        ArrayList<Integer> sim = new ArrayList<>(board);
        sim.set(index, newVal);
        for (int i = 0; i < queens; i++) {
            for (int j = i+1; j < queens; j++) {
                if (sim.get(i).equals(sim.get(j)) || (Math.abs(sim.get(j) - sim.get(i)) == j - i)) res++;
            }
        }
        return res;
    }


    public MARCState getRandomState() {
        Random r = new Random();
        ArrayList<Integer> board = new ArrayList<>();
        for (int i = 0; i < queens; i++) board.add(r.nextInt(queens));
        return new QueensBoard(this, board);
    }

    public Integer getDim() {
        return this.queens;
    }

    public NQueens(Integer dim) {
        super(dim);
        this.queens = dim;
    }


}

